### PR TITLE
gRPC - archive with a MutinyBean implementor is a bean archive

### PR DESCRIPTION
**Pull request by mkouba**

### CODE REVIEW

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/BeanArchivePredicateBuildItem.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+import io.quarkus.builder.item.MultiBuildItem;
+import io.quarkus.deployment.ApplicationArchive;
+
+/**
+ * 
+ * By default, only explict/implicit bean archives (as defined by the spec) are considered during the bean discovery. However,
+ * extensions can register a logic to identify additional bean archives.
+ */
+public final class BeanArchivePredicateBuildItem extends MultiBuildItem {
```
Comment (mkouba) [inline extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/BeanArchivePredicateBuildItem.java:13]: I'm not sure about the name of this build item. The other alternative was `AdditionalBeanArchiveBuildItem` but given the fact that the item does not represent a bean archive but a logic that identifies a bean archive I decided for `BeanArchivePredicateBuildItem`.

Comment (Ladicek) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/BeanArchivePredicateBuildItem.java:13]: I think `BeanArchivePredicateBuildItem` makes sense.

**Review by Ladicek — APPROVED** [review]
Test for the original issue would be nice, but I see how that's not straightforward to add.

**Comment by mkouba** [comment]
> Test for the original issue would be nice, but I see how that's not straightforward to add.

I don't know of a way how to test it TBH...

**Review by manovotn — APPROVED** [review]

File: extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java [hunk 2]
```diff
@@ -546,6 +548,18 @@ void configureMetrics(GrpcBuildTimeConfig configuration, Optional<MetricsCapabil
...
     }
 
+    @BuildStep
+    BeanArchivePredicateBuildItem additionalBeanArchives() {
+        return new BeanArchivePredicateBuildItem(new Predicate<ApplicationArchive>() {
+
+            @Override
+            public boolean test(ApplicationArchive archive) {
+                // Every archive that contains a generated implementor of MutinyBean is considered a bean archive
+                return !archive.getIndex().getKnownDirectImplementors(GrpcDotNames.MUTINY_BEAN).isEmpty();
```
Comment (michalszynkiewicz) [inline extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java:558]: is `archive.getIndex()` always `!= null`?

**Comment by michalszynkiewicz** [comment]
nice! Thanks Martin!

File: extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java (hunk 2 above)
Comment (mkouba) [reply extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java:558]: I believe so. But it's a good point that we could be more defensive here...

Comment (mkouba) [reply extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java:558]: Hm, we don't test the nullability in the `BeanArchiveProcessor` either. So it's probably OK.

**Review by michalszynkiewicz — APPROVED** [review]

**Comment by mkouba** [comment]
Windows CI failures do not seem to be related.

**Pull request merged**